Source/WebKit/UIProcess/mac/DisplayLinkProcessProxyClient.cpp

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "config.h"
+#include "DisplayLinkProcessProxyClient.h"
+
+#if HAVE(CVDISPLAYLINK)
+
+#include "EventDispatcherMessages.h"
+#include "WebProcessMessages.h"
+
+namespace WebKit {
+
+void DisplayLinkProcessProxyClient::setConnectionID(IPC::Connection::UniqueID connectionID)
+{
+    m_connectionID = connectionID;
+}
+
+// This is called off the main thread.
+void DisplayLinkProcessProxyClient::displayLinkFired(WebCore::PlatformDisplayID displayID, WebCore::DisplayUpdate displayUpdate, bool wantsFullSpeedUpdates, bool anyObserverWantsCallback)
+{
+    IPC::Connection::UniqueID connectionID = m_connectionID;
+    if (!connectionID)
+        return;
+
+    if (wantsFullSpeedUpdates)
+        IPC::Connection::send(connectionID, Messages::EventDispatcher::DisplayWasRefreshed(displayID, displayUpdate, anyObserverWantsCallback), 0, { }, Thread::QOS::UserInteractive);
```

Comment:
I know it's an existing problem, but:
send via `UniqueID` is not thread-safe. It will potentially revive a 0 refcount object and cause UAF + double deletion.
Instead of storing `UniqueID`, `RefPtr<Connection>` should be used, as is business-as-usual with refcounted objects in general.

Reply:
I cache an IPC::Connection now.